File: lib/extractor.py
import abc
import re
import requests
import json

from bs4 import BeautifulSoup


class InfoExtractor(abc.ABC):
    @abc.abstractmethod
    def extract(video_url):
        pass


class PlaylistInfoExtractor(InfoExtractor):
    @staticmethod
    def extract(video_url):
        try:
            response = requests.get(video_url)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')
            script_tags = soup.find_all('script')
            for script_tag in script_tags:
                if 'ytInitialData' in script_tag.text:
                    script_content = script_tag.get_text()
                    match = re.search(r'var\s+ytInitialData\s*=\s*({.*?});',
                                      script_content)
                    if match:
                        yt_initial_data_str = match.group(1)
                        data = json.loads(yt_initial_data_str)
                        album_name = (
                            data['metadata']['playlistMetadataRenderer']
                            ['albumName'])
                        song_urls = []
                        for d in (
                            data['contents']['twoColumnBrowseResultsRenderer']['tabs'][0][
                                'tabRenderer']['content']['sectionListRenderer']['contents'][0][
                                    'itemSectionRenderer']['contents'][0]['playlistVideoListRenderer'][
                                        'contents']
                                ):
                            if 'playlistVideoRenderer' in d:
                                song_urls.append(d['playlistVideoRenderer']['navigationEndpoint'][
                                    'commandMetadata']['webCommandMetadata']['url'])
                        return album_name, [
                            f'https://www.youtube.com{url}' for url in song_urls]
            raise Exception("Not found")
        except Exception as e:
            print(f"Error extracting youtube playlist info: {e}")
            return None


class SongInfoExtractor(InfoExtractor):
    @classmethod
    def extract(cls, video_url):
        try:
            response = requests.get(video_url)
            response.raise_for_status()
            soup = BeautifulSoup(response.text, 'html.parser')
            script_tags = soup.find_all('script')
            for script_tag in script_tags:
                if 'ytInitialData' in script_tag.text:
                    script_content = script_tag.get_text()
                    match = re.search(r'var\s+ytInitialData\s*=\s*({.*?});',
                                      script_content)
                    if match:
                        yt_initial_data_str = match.group(1)
                        data = json.loads(yt_initial_data_str)
                        for panel in data.get('engagementPanels', []):
                            if 'engagementPanelSectionListRenderer' in panel:
                                panel_data = panel['engagementPanelSectionListRenderer']
                                content_list = panel_data['content']
                                if 'structuredDescriptionContentRenderer' in content_list:
                                    item_data = content_list[
                                        'structuredDescriptionContentRenderer']
                                    cards = item_data.get('items', [])
                                    for card in cards:
                                        if 'horizontalCardListRenderer' in card:
                                            return cls._extract_song(card)
                                        if 'videoDescriptionHeaderRenderer' in card:
                                            return cls._extract_video(card, video_url)
                                        raise Exception('Data cannot be extracted')
            raise Exception("Not found")
        except Exception as e:
            print(f"Error extracting youtube song info: {e}")
            return None

    def _extract_song(card):
        card_data = card['horizontalCardListRenderer']
        cards_in_list = card_data.get('cards', [])
        for card_item in cards_in_list:
            if 'videoAttributeViewModel' in card_item:
                data = card_item['videoAttributeViewModel']
                song_title = data['title']
                artist_name = data['subtitle']
                album_name = data['secondarySubtitle']['content']
                cover_img_url = data['image']['sources'][0]['url']
                return song_title, artist_name, album_name, cover_img_url

    def _extract_video(card, video_url):
        card_data = card['videoDescriptionHeaderRenderer']
        song_title = card_data['title']['runs'][0]['text']
        artist_name = card_data['channel']['simpleText']
        album_name = song_title
        return song_title, artist_name, album_name, video_url

File: lib/extractor.rb
require 'net/http'
require 'nokogiri'
require 'json'

module InfoExtractor
  def extract(video_url)
    raise NotImplementedError, "#{self.class.name}##{__method__} is an abstract method."
  end
end

class PlaylistInfoExtractor extend InfoExtractor
  def self.extract(video_url)
    uri = URI(video_url)
    response = Net::HTTP.get_response(uri)
    raise "HTTP Error: #{response.code}" unless response.is_a?(Net::HTTPSuccess)

    soup = Nokogiri::HTML(response.body.force_encoding('UTF-8'))
    script_tags = soup.css('script')
    script_tags.each do |script_tag|
      if script_tag.content.include?('ytInitialData')
        match = script_tag.content.match(/var\s+ytInitialData\s*=\s*(\{.*?\});/)
        if match
          yt_initial_data = JSON.parse(match[1])
          album_name = yt_initial_data.dig('metadata', 'playlistMetadataRenderer', 'albumName')
          song_urls = yt_initial_data.dig('contents', 'twoColumnBrowseResultsRenderer', 'tabs', 0, 'tabRenderer', 'content', 'sectionListRenderer', 'contents', 0, 'itemSectionRenderer', 'contents', 0, 'playlistVideoListRenderer', 'contents').select { |content| content.key?('playlistVideoRenderer') }.map { |vid| "https://www.youtube.com" + vid.dig('playlistVideoRenderer', 'navigationEndpoint', 'commandMetadata', 'webCommandMetadata', 'url') }
          return album_name, song_urls
        end
      end
    end
    raise "Playlist not found"
  rescue => e
    puts "Error extracting Youtube playlist info: #{e}"
    nil
  end
end


class SongInfoExtractor extend InfoExtractor
  def self.extract(video_url)
    uri = URI(video_url)
    response = Net::HTTP.get_response(uri)
    raise "HTTP ERROR: #{response.code}" unless response.is_a?(Net::HTTPSuccess)

    soup = Nokogiri::HTML(response.body.force_encoding('UTF-8'))
    script_tags = soup.css('script')
    script_tags.each do |script_tag|
      if script_tag.content.include?('ytInitialData')
        match = script_tag.content.match(/var\s+ytInitialData\s*=\s*(\{.*?\});/)
        if match
          yt_initial_data = JSON.parse(match[1])
          yt_initial_data['engagementPanels'].each do |panel|

            if panel['engagementPanelSectionListRenderer']
              item_data = panel['engagementPanelSectionListRenderer']&.dig('content', 'structuredDescriptionContentRenderer', 'items')
              if item_data
                item_data.each do |card|
                  if card['horizontalCardListRenderer']
                    return extract_song(card)
                  end
                end
                item_data.each do |card|
                  if card['videoDescriptionHeaderRenderer']
                    return extract_video(card, video_url)
                  end
                end
              end
            end
          end
        end
      end
    end
    raise "Not found"
    rescue => e
      puts "Error extracting youtube song info: #{e}"
    nil
  end

  def self.extract_song(card)
    card_data = card['horizontalCardListRenderer']['cards']
    card_data.each do |card_item|
      if card_item['videoAttributeViewModel']
        data = card_item['videoAttributeViewModel']
        song_title = data['title'].encode('UTF-8')
        artist_name = data['subtitle'].encode('UTF-8')
        album_name = data['secondarySubtitle']
        cover_img_url = data['image']['sources'][0]['url']
        return song_title, artist_name, album_name, cover_img_url
      end
    end
  end

  def self.extract_video(card, video_url)
    card_data = card['videoDescriptionHeaderRenderer']
    song_title = card_data['title']['runs'][0]['text']
    artist_name = card_data['channel']['simpleText']
    album_name = song_title
    return song_title, artist_name, album_name, video_url
  end
end
